sdk/identity/azure-identity/azure/identity/_credentials/certificate.py

```diff
@@ -76,6 +77,42 @@ def extract_cert_chain(pem_bytes):
     return b"".join(chain.splitlines())
 
 
+_Cert = NamedTuple("_Cert", [("pem_bytes", bytes), ("private_key", "Any"), ("fingerprint", bytes)])
+
+
+def load_pem_certificate(certificate_data, password):
+    # type: (bytes, Optional[bytes]) -> _Cert
+    private_key = serialization.load_pem_private_key(certificate_data, password, backend=default_backend())
+    cert = x509.load_pem_x509_certificate(certificate_data, default_backend())
+    fingerprint = cert.fingerprint(hashes.SHA1())  # nosec
+    return _Cert(certificate_data, private_key, fingerprint)
+
+
+def load_pkcs12_certificate(certificate_data, password):
+    # type: (bytes, Optional[bytes]) -> _Cert
+    from cryptography.hazmat.primitives.serialization import Encoding, NoEncryption, pkcs12, PrivateFormat
+
+    private_key, cert, additional_certs = pkcs12.load_key_and_certificates(
+        certificate_data, password, backend=default_backend()
+    )
+    if not private_key:
+        raise ValueError("The certificate must include its private key")
+    if not cert:
+        # mentioning PEM here because we raise this error when certificate_data is garbage
+        raise ValueError("Failed to deserialize certificate in PEM or PKCS12 format")
+
+    # This serializes the private key without any encryption it may have had. Doing so doesn't violate security
+    # boundaries because this representation of the key is kept in memory. We already have the key and its
+    # password, if any, in memory.
+    key_bytes = private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8, NoEncryption())
+    pem_sections = [key_bytes] + [c.public_bytes(Encoding.PEM) for c in [cert] + additional_certs]
+    pem_bytes = b"".join(pem_sections)
+
+    fingerprint = cert.fingerprint(hashes.SHA1())  # nosec
+
+    return _Cert(pem_bytes, private_key, fingerprint)
+
+
 def get_client_credential(certificate_path, password=None, certificate_data=None, send_certificate_chain=False, **_):
```

Comment:
Seems like certificate_path is not required. Why not give it a default value?

Reply:
Either certificate_path or certificate_data is required. `CertificateCredential.__init__` will always have a value for certificate_path because it was originally a required parameter.

Reply:
Then if user wants to use certificate_data, they have to explicitly set certificate_path to None?

My concern (though not very strong) is if we imply user must set the value of certificate_path and if certificate_path is not None, setting certificate_data raises ValueError. It will cause confusion if user wants to use certificate_data to get_client_credential?

Reply:
This (get_client_credential) is an internal helper method called only by `CertificateCredential.__init__`, so I doubt it implies anything to users. This is the public API:
```py
# load a cert from disk
CertificateCredential(tenant_id, client_id, "/cert.pem")

# load a cert from bytes
CertificateCredential(tenant_id, client_id, certificate_data=some_bytes)
```

Reply:
Get it. Then never mind.

Thanks for the clarification.